django/db/models/sql/datastructures.py

```diff
@@ -114,17 +114,28 @@ def relabeled_clone(self, change_map):
             self.join_field, self.nullable, filtered_relation=filtered_relation,
         )
 
-    def equals(self, other, with_filtered_relation):
+    @property
+    def identity(self):
         return (
-            isinstance(other, self.__class__) and
-            self.table_name == other.table_name and
-            self.parent_alias == other.parent_alias and
-            self.join_field == other.join_field and
-            (not with_filtered_relation or self.filtered_relation == other.filtered_relation)
+            self.__class__,
+            self.table_name,
+            self.parent_alias,
+            self.join_field,
+            self.filtered_relation,
         )
 
     def __eq__(self, other):
-        return self.equals(other, with_filtered_relation=True)
+        if not isinstance(other, Join):
```

Comment:
Is there a particular reason not to use `self.__class__`?
```suggestion
        if not isinstance(other, self.__class__):
```

Reply:
yes, because the following comparison is only valid for `Join` instances while `self.__class__` could be restricting if we ever add a `Join` subclass.